File: src/main/java/com/zhongan/devpilot/completions/common/completions/CompletionUtils.java
package com.zhongan.devpilot.completions.common.completions;

import static com.zhongan.devpilot.completions.common.general.StaticConfig.MAX_COMPLETIONS;
import static com.zhongan.devpilot.completions.common.general.Utils.endsWithADot;

import com.intellij.codeInsight.completion.CompletionParameters;
import com.intellij.codeInsight.completion.CompletionResultSet;
import com.intellij.openapi.diagnostic.Logger;
import com.intellij.openapi.editor.Document;
import com.intellij.openapi.util.TextRange;
import com.zhongan.devpilot.completions.common.requests.ResultEntry;
import com.zhongan.devpilot.completions.common.general.SuggestionTrigger;
import com.zhongan.devpilot.completions.common.prediction.DevPilotCompletion;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class CompletionUtils {

    @Nullable
    private static String getCursorPrefix(@NotNull Document document, int cursorPosition) {
        try {
            int lineNumber = document.getLineNumber(cursorPosition);
            int lineStart = document.getLineStartOffset(lineNumber);

            return document.getText(TextRange.create(lineStart, cursorPosition)).trim();
        } catch (Throwable e) {
            Logger.getInstance(CompletionUtils.class).warn("Failed to get cursor prefix: ", e);
            return null;
        }
    }

    @Nullable
    private static String getCursorSuffix(@NotNull Document document, int cursorPosition) {
        try {
            int lineNumber = document.getLineNumber(cursorPosition);
            int lineEnd = document.getLineEndOffset(lineNumber);

            return document.getText(TextRange.create(cursorPosition, lineEnd)).trim();
        } catch (Throwable e) {
            Logger.getInstance(CompletionUtils.class).warn("Failed to get cursor suffix: ", e);
            return null;
        }
    }

    @Nullable
    public static DevPilotCompletion createDevpilotCompletion(
            @NotNull Document document,
            int offset,
            String oldPrefix,
            ResultEntry result,
            int index,
            SuggestionTrigger suggestionTrigger) {
        String cursorPrefix = CompletionUtils.getCursorPrefix(document, offset);
        String cursorSuffix = CompletionUtils.getCursorSuffix(document, offset);
        if (cursorPrefix == null || cursorSuffix == null) {
            return null;
        }

        return new DevPilotCompletion(
                oldPrefix,
                result.new_prefix,
                result.old_suffix,
                result.new_suffix,
                index,
                cursorPrefix,
                cursorSuffix,
                result.completion_metadata,
                suggestionTrigger);
    }

    public static int completionLimit(
            CompletionParameters parameters, CompletionResultSet result, boolean isLocked) {
        return completionLimit(
                parameters.getEditor().getDocument(),
                result.getPrefixMatcher().getPrefix(),
                parameters.getOffset(),
                isLocked);
    }

    public static int completionLimit(
            @NotNull Document document, @NotNull String prefix, int offset, boolean isLocked) {
        if (isLocked) {
            return 1;
        }
        boolean preferDevPilot = !endsWithADot(document, offset - prefix.length());

        return preferDevPilot ? MAX_COMPLETIONS : 1;
    }
}

File: src/main/java/com/zhongan/devpilot/completions/inline/CompletionUtils.kt
package com.zhongan.devpilot.completions.inline

import com.intellij.openapi.editor.Document
import com.intellij.openapi.util.TextRange
import java.util.regex.Pattern

object CompletionUtils {
    private val END_OF_LINE_VALID_PATTERN = Pattern.compile("^\\s*[)}\\]\"'`]*\\s*[:{;,]?\\s*$")

    @JvmStatic
    fun isValidDocumentChange(document: Document, newOffset: Int, previousOffset: Int): Boolean {
        if (newOffset < 0 || previousOffset > newOffset) return false

        val addedText = document.getText(TextRange(previousOffset, newOffset))
        return (
                isValidMidlinePosition(document, newOffset) &&
                        isValidNonEmptyChange(addedText.length, addedText) &&
                        isSingleCharNonWhitespaceChange(addedText)
                )
    }

    @JvmStatic
    fun isValidMidlinePosition(document: Document, offset: Int): Boolean {
        val lineIndex: Int = document.getLineNumber(offset)
        val lineRange = TextRange.create(document.getLineStartOffset(lineIndex), document.getLineEndOffset(lineIndex))
        val line = document.getText(lineRange)
        val lineSuffix = line.substring(offset - lineRange.startOffset)
        return END_OF_LINE_VALID_PATTERN.matcher(lineSuffix).matches()
    }

    @JvmStatic
    fun isValidNonEmptyChange(replacedTextLength: Int, newText: String): Boolean {
        return replacedTextLength >= 0 && newText != ""
    }

    @JvmStatic
    fun isSingleCharNonWhitespaceChange(newText: String): Boolean {
        return newText.trim().length <= 1
    }
}
